### PR TITLE
[rb] Add all the RBS files to add full rbs support to all selenium rb classes and modules

### PR DESCRIPTION
**Pull request by aguspe**
### Description
This PR adds more than 120 RBS files and updates the steep file to reduce all the current steep errors.
The goal with this PR is that afterward it would be possible to add a steep check on the bazel pipeline.

### Motivation and Context
Based on #10943 I created the following PRs to add RBS type support to the ruby selenium library:

On #12844 I started adding support for RBS and I added the steep file configuration
On #13192 I extended the classes supported on RBS and updated the logger file

I stated on #13192 that I wanted to make smaller PRs to expand support and make it easier to review, but since there was a lot of work to be done to cover all the classes and modules in the library I decided to expand this PR to add all the missing coverage and fix all the steep issues

Right now if a steep check is run after the two previously mentioned PRs this is the result:

Screenshot 2023-12-02 at 20 31 30

Already with the files added and the extension of libraries on the steep config file the amount of errors drops to:

Screenshot 2023-12-02 at 19 57 48

This is currently a work in progress and I will submit this PR after all the type errors are reduced to 0 and all the files are properly updated to reflect the right types used in classes and not just undefined

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

### Checklist

- [x] I have read the [contributing](https://github.com/SeleniumHQ/selenium/blob/trunk/CONTRIBUTING.md) document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by aguspe** [comment]
Screenshot 2024-02-26 at 20 45 19

These are the errors left from the original 1558, I will put it on ready for review, because it will probably take a while for review and it will need more updates

Also all the unit tests are passing @titusfortner

**Comment by aguspe** [comment]
The latest update after updating the ignore list, I wrote in the steep file the 62 ignore errors across 28 files, I will keep investigating in the case of overloading issues or libraries such as URI and NET::HTTP that give unwanted RBS issues:

Screenshot 2024-03-06 at 22 31 20

Then of course we will need to keep updating the files to replace all the untype signatures where is fit

**Comment by aguspe** [comment]
Hi @p0deje this is ready for review, please let me know if something is missing or if I can do anything to facilitate the review, in case this PR is not relevant anymore it can also be closed, thank you and have a great day

File: rb/Gemfile.lock [hunk 1]
```diff
@@ -51,11 +51,9 @@ GEM
     i18n (1.14.1)
       concurrent-ruby (~> 1.0)
     io-console (0.7.2)
-    io-console (0.7.2-java)
```
Comment (p0deje) [inline rb/Gemfile.lock]: This would need to be reverted to make sure the CI passes on JRuby.

File: rb/Steepfile [hunk 2]
```diff
@@ -1,8 +1,76 @@
-# frozen_string_literal: true
-
 target :lib do
-  signature 'sig' # Signature directory
+  signature 'sig', '.gem_rbs_collection/rubyzip' # Signature directory
   check 'lib' # Directory name
+  # Total amount of errors ignore 61 in 27 files
+  ignore(
+    # Ignore line 166 due to UDP RBS issue
+    'lib/selenium/webdriver/common/platform.rb',
```
Comment (p0deje) [inline rb/Steepfile:10]: Is it possible to disable steep inside the file itself? Something similar to `# rubocop:disable` comment.

File: rb/lib/selenium/devtools/support/cdp_client_generator.rb [hunk 3]
```diff
@@ -26,8 +26,9 @@ module DevTools
     module Support
       class CDPClientGenerator
         # Input JSON files are generated from PDL tasks.
-        DOMAIN_TEMPLATE_PATH = File.expand_path('cdp/domain.rb.erb', __dir__)
-        LOADER_TEMPLATE_PATH = File.expand_path('cdp/loader.rb.erb', __dir__)
+        DIR = __dir__ || ''
```
Comment (p0deje) [inline rb/lib/selenium/devtools/support/cdp_client_generator.rb:29]: Why is this necessary?

File: rb/lib/selenium/webdriver/common/driver.rb [hunk 4]
```diff
@@ -71,7 +71,7 @@ def for(browser, opts = {})
       def initialize(bridge: nil, listener: nil, **opts)
         @devtools = nil
         @bidi = nil
-        bridge ||= create_bridge(**opts)
+        bridge ||= create_bridge(caps: opts[:caps], url: opts[:url], http_client: opts[:http_client])
```
Comment (p0deje) [inline rb/lib/selenium/webdriver/common/driver.rb]: I feel like there could be more in `opts` than listed. Is there any way to keep using kwargs?

File: rb/lib/selenium/webdriver/common/manager.rb [hunk 5]
```diff
@@ -56,7 +56,7 @@ def add_cookie(opts = {})
         opts[:httpOnly] = http_only if http_only
 
         obj = opts.delete(:expires)
-        opts[:expiry] = seconds_from(obj).to_i if obj
+        opts[:expiry] = seconds_from(obj).to_int if obj
```
Comment (p0deje) [inline rb/lib/selenium/webdriver/common/manager.rb:59]: What's the difference between to_i and to_int? Should we instead use `Integer()`?

File: rb/lib/selenium/webdriver/firefox/driver.rb [hunk 6]
```diff
@@ -55,6 +55,7 @@ def devtools_url
           end
 
           uri = URI("http://#{capabilities['moz:debuggerAddress']}")
+
```
Comment (p0deje) [inline rb/lib/selenium/webdriver/firefox/driver.rb]: This is not necessary, right?

File: rb/lib/selenium/webdriver/firefox/profile.rb [hunk 7]
```diff
@@ -24,6 +24,7 @@ class Profile
         include ProfileHelper
 
         VALID_PREFERENCE_TYPES = [TrueClass, FalseClass, Integer, Float, String].freeze
+        WEBDRIVER_PREFS = []
```
Comment (p0deje) [inline rb/lib/selenium/webdriver/firefox/profile.rb]: Do we need this constant? I don't see where it's used.

File: rb/lib/selenium/webdriver/safari/options.rb [hunk 8]
```diff
@@ -22,8 +22,6 @@ module WebDriver
     module Safari
       class Options < WebDriver::Options
         attr_accessor :options
-
-        # @see https://developer.apple.com/documentation/webkit/about_webdriver_for_safari
```
Comment (p0deje) [inline rb/lib/selenium/webdriver/safari/options.rb]: Please revert.

File: rb/Steepfile [hunk 9]
```diff
@@ -1,8 +1,76 @@
-# frozen_string_literal: true
-
 target :lib do
-  signature 'sig' # Signature directory
+  signature 'sig', '.gem_rbs_collection/rubyzip' # Signature directory
```
Comment (p0deje) [inline rb/Steepfile:5]: I don't see rubyzip committed, is this by mistake?

File: rb/sig/gems/open3/open3.rbs [hunk 10]
```diff
@@ -0,0 +1,3 @@
+module Open3
```
Comment (p0deje) [inline rb/sig/gems/open3/open3.rbs:1]: Do we have to commit all the gems' signatures to the repository? I wonder if we can just pull them up in the initialization phase.

File: rb/Steepfile (hunk 9 above)
Comment (aguspe) [reply rb/Steepfile:5]: @p0deje Without specifying rubyzip like this, we get the following error, because we have ruby zip as a dependency

So in our rbs_collection.yaml, when we run rbs collection install we get all the signatures we need

Here is the reference: https://github.com/ruby/gem_rbs_collection

Screenshot 2024-03-13 at 21 42 04

Screenshot 2024-03-13 at 22 03 29

File: rb/sig/gems/open3/open3.rbs (hunk 10 above)
Comment (aguspe) [reply rb/sig/gems/open3/open3.rbs:1]: This is possible, but the reason why I did this was because gems such as:

https://github.com/ruby/open3

Do not have RBS signatures, so I could add the whole signature for the gems we need https://github.com/ruby/gem_rbs_collection, and it will be nice for the community but it will also delayed the PR because it will be the same type of work here, but for all the dependent gems

So I just did the minimum to get unblocked by other gems lacking signatures

File: rb/Steepfile (hunk 2 above)
Comment (aguspe) [reply rb/Steepfile:10]: I haven't found anything online but I asked on the discord server, if I get the confirmation that is not possible I will create a PR in the steep project because it will be useful

File: rb/lib/selenium/webdriver/common/manager.rb (hunk 5 above)
Comment (aguspe) [reply rb/lib/selenium/webdriver/common/manager.rb:59]: In this scenario the challenge is that seconds_from can return a Numeric type and the Numeric class only has to_int

RBS signature for seconds_from:

`def seconds_from: (Time | DateTime | Numeric obj) -> (Float | Numeric)`

References for Numeric:

https://ruby-doc.org/core-2.5.1/Numeric.html

Here is a nice article on also different usages:

https://www.rubyfleebie.com/2007/04/05/to_i-vs-to_int/

Error with to_i:

Screenshot 2024-03-13 at 22 21 24

File: rb/lib/selenium/webdriver/firefox/profile.rb (hunk 7 above)
Comment (aguspe) [reply rb/lib/selenium/webdriver/firefox/profile.rb]: So I was trying to solve this issue because this contact appears here, and throws me the following validation error:

Screenshot 2024-03-13 at 22 22 03

File: rb/lib/selenium/devtools/support/cdp_client_generator.rb (hunk 3 above)
Comment (aguspe) [reply rb/lib/selenium/devtools/support/cdp_client_generator.rb:29]: When using __dir__ directly it's possible to return nil, and that produces and error type since the expectation for the file library is a string or an object that can receive _ToPath

Screenshot 2024-03-13 at 22 26 20
Screenshot 2024-03-13 at 22 25 56

File: rb/lib/selenium/webdriver/common/driver.rb (hunk 4 above)
Comment (aguspe) [reply rb/lib/selenium/webdriver/common/driver.rb]: I agree, I tried different variations in the signature and to overload it but I kept getting the same error,
this is why several methods that use kwargs are ignored in the steepfile, I found the same issue here:

https://github.com/soutaro/steep/issues/429

And I don't think the fix is actually working, I will roll this back and add the file to ignore as with the other kwargs issues

File: rb/Gemfile.lock (hunk 1 above)
Comment (aguspe) [reply rb/Gemfile.lock]: Reverted

File: rb/lib/selenium/webdriver/firefox/driver.rb (hunk 6 above)
Comment (aguspe) [reply rb/lib/selenium/webdriver/firefox/driver.rb]: No thank you for spotting it, reverted

File: rb/lib/selenium/webdriver/safari/options.rb (hunk 8 above)
Comment (aguspe) [reply rb/lib/selenium/webdriver/safari/options.rb]: Reverted

File: rb/Steepfile (hunk 2 above)
Comment (aguspe) [reply rb/Steepfile:10]: I got some help on Discord and there is a merged PR for the release 1.7.0 in steep that adds support for this, but the release is not ready yet, so for now this is the only way to ignore

https://github.com/soutaro/steep/pull/1034

**Comment by aguspe** [comment]
Thank you for the quick review and the help @p0deje, here in Denmark is quite late now, so I will check all the answers tomorrow morning

File: rb/Steepfile (hunk 2 above)
Comment (p0deje) [reply rb/Steepfile:10]: We could just install steep from GitHub via Gemfile directly to avoid waiting for release.

```ruby
gem 'steep', github: 'soutaro/steep', require: false, platforms: %i[mri mingw x64_mingw]
```

File: rb/lib/selenium/devtools/support/cdp_client_generator.rb (hunk 3 above)
Comment (p0deje) [reply rb/lib/selenium/devtools/support/cdp_client_generator.rb:29]: That's unfortunate, though I would never expect `__dir__` to return nil when called from a script.

File: rb/lib/selenium/webdriver/common/driver.rb (hunk 4 above)
Comment (p0deje) [reply rb/lib/selenium/webdriver/common/driver.rb]: Should we define the signature as?

```ruby
def create_bridge: (Hash[Symbol, String]) -> untyped
```

File: rb/lib/selenium/webdriver/common/manager.rb (hunk 5 above)
Comment (p0deje) [reply rb/lib/selenium/webdriver/common/manager.rb:59]: Interesting, TIL.

File: rb/lib/selenium/webdriver/firefox/profile.rb (hunk 7 above)
Comment (p0deje) [reply rb/lib/selenium/webdriver/firefox/profile.rb]: This is a great catch, looks like the constant was lost somewhere during refactorings, but looked like this https://github.com/SeleniumHQ/selenium/blob/560cb6b427780542d7dcd722ab6fc09ff21e65b3/rb/lib/selenium/webdriver/firefox/profile.rb#L9-L15. Let's bring it back at least with `port` and `log_file`.

File: rb/sig/gems/open3/open3.rbs (hunk 10 above)
Comment (p0deje) [reply rb/sig/gems/open3/open3.rbs:1]: Ok, noted, thanks for doing that!

File: rb/Steepfile (hunk 9 above)
Comment (p0deje) [reply rb/Steepfile:5]: I don't quite follow what populates `.gem_rbs_collection/rubyzip' - I cannot find the rbs_collection.yaml entry for RubyZip. I would probably need to play around with this PR to understand better.

File: rb/lib/selenium/webdriver/firefox/profile.rb (hunk 7 above)
Comment (aguspe) [reply rb/lib/selenium/webdriver/firefox/profile.rb]: I added log_file & port back

File: rb/lib/selenium/webdriver/common/driver.rb (hunk 4 above)
Comment (aguspe) [reply rb/lib/selenium/webdriver/common/driver.rb]: The issue is that by doing that, the other uses of create_bridge, I get a MethodArityMismatch error and MethodParameterMismatch

it's the same if I put both values as nillable or if I overload the method declaration

I need to investigate more on the kwargs implementation because is not the only place that I have this issue

Screenshot 2024-03-16 at 00 28 28

File: rb/Steepfile (hunk 9 above)
Comment (aguspe) [reply rb/Steepfile:5]: The rubyzip signatures are in the .gem_rbs_collection, so when we get all the signatures after doing rbs install it gets all the signatures from the gem_rbs_collection repo

Screenshot 2024-03-16 at 00 33 29
Screenshot 2024-03-16 at 00 33 52

File: rb/Steepfile (hunk 2 above)
Comment (aguspe) [reply rb/Steepfile:10]: I tried this, and I followed up the new way to ignore per line, but I cannot make it work, I will try tomorrow and reach out again in discord, but it seemed with the steep release 1.7.0-dev2 I couldn't make it work

Comment (aguspe) [reply rb/Steepfile:10]: I will make a PR updating the ignore files when release 1.7.0 is released and add the lines that should be ignored

**Comment by aguspe** [comment]
I made one more commit to remove all the comments created during auto generation, and cast the generic type to the constants that were mapped to a specific string or symbol during autogeneration, also I added some refactoring in general

After the last steep check, there are still no errors, and from there we can start replacing the untyped values for the right values and correct each error one at the time

@p0deje 

Screenshot 2024-03-17 at 00 33 39

**Review by p0deje — APPROVED** [review]

**Comment by p0deje** [comment]
Can you please fix RuboCop complaints? 
```
Inspecting 276 files
....................................................................................................................................CC...............C..............................................................................................................................

Offenses:

rb/lib/selenium/webdriver/firefox/profile.rb:27:27: C: [Correctable] Style/MutableConstant: Freeze mutable objects assigned to constants.
        WEBDRIVER_PREFS = { ...
                          ^
rb/lib/selenium/webdriver/firefox/profile.rb:28:11: C: [Corrected] Style/HashSyntax: Use the new Ruby 1.9 hash syntax.
          :port => 'webdriver_firefox_port',
          ^^^^^^^^
rb/lib/selenium/webdriver/firefox/profile.rb:29:11: C: [Corrected] Style/HashSyntax: Use the new Ruby 1.9 hash syntax.
          :log_file => 'webdriver.log.file'
          ^^^^^^^^^^^^
rb/lib/selenium/webdriver/firefox/profile.rb:169:37: C: [Corrected] Layout/SpaceInsideHashLiteralBraces: Space inside { detected.
            WebDriver.logger.debug({ extension: name }.inspect, id: :firefox_profile)
                                    ^
rb/lib/selenium/webdriver/firefox/profile.rb:169:53: C: [Corrected] Layout/SpaceInsideHashLiteralBraces: Space inside } detected.
            WebDriver.logger.debug({ extension: name }.inspect, id: :firefox_profile)
                                                    ^
rb/lib/selenium/webdriver/firefox/profiles_ini.rb:44:9: C: Metrics/CyclomaticComplexity: Cyclomatic complexity for parse is too high. [11/9]
        def parse ...
        ^^^^^^^^^
rb/lib/selenium/webdriver/remote/http/default.rb:70:11: C: Metrics/AbcSize: Assignment Branch Condition size for request is too high. [<5, 29, 8> 30.5/30]
          def request(verb, url, headers, payload, redirects = 0) ...
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

**Comment by aguspe** [comment]
> Can you please fix RuboCop complaints?
> 
> ```
> Inspecting 276 files
> ....................................................................................................................................CC...............C..............................................................................................................................
> 
> Offenses:
> 
> rb/lib/selenium/webdriver/firefox/profile.rb:27:27: C: [Correctable] Style/MutableConstant: Freeze mutable objects assigned to constants.
>         WEBDRIVER_PREFS = { ...
>                           ^
> rb/lib/selenium/webdriver/firefox/profile.rb:28:11: C: [Corrected] Style/HashSyntax: Use the new Ruby 1.9 hash syntax.
>           :port => 'webdriver_firefox_port',
>           ^^^^^^^^
> rb/lib/selenium/webdriver/firefox/profile.rb:29:11: C: [Corrected] Style/HashSyntax: Use the new Ruby 1.9 hash syntax.
>           :log_file => 'webdriver.log.file'
>           ^^^^^^^^^^^^
> rb/lib/selenium/webdriver/firefox/profile.rb:169:37: C: [Corrected] Layout/SpaceInsideHashLiteralBraces: Space inside { detected.
>             WebDriver.logger.debug({ extension: name }.inspect, id: :firefox_profile)
>                                     ^
> rb/lib/selenium/webdriver/firefox/profile.rb:169:53: C: [Corrected] Layout/SpaceInsideHashLiteralBraces: Space inside } detected.
>             WebDriver.logger.debug({ extension: name }.inspect, id: :firefox_profile)
>                                                     ^
> rb/lib/selenium/webdriver/firefox/profiles_ini.rb:44:9: C: Metrics/CyclomaticComplexity: Cyclomatic complexity for parse is too high. [11/9]
>         def parse ...
>         ^^^^^^^^^
> rb/lib/selenium/webdriver/remote/http/default.rb:70:11: C: Metrics/AbcSize: Assignment Branch Condition size for request is too high. [<5, 29, 8> 30.5/30]
>           def request(verb, url, headers, payload, redirects = 0) ...
>           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> ```

Oh sorry for that, I forgot to run rubocop locally, my bad

Here is the latest rubocop and steep check run all with 0 issues, I added a rubocop ignore on the steep file because the steep config file doesn't follow rubocop recommended block size

Screenshot 2024-03-17 at 17 56 18
Screenshot 2024-03-17 at 17 56 05

**Comment by p0deje** [comment]
Out of curiosity, what are these warnings?

> [#type_check_file(lib/selenium/webdriver/remote/http/common.rb@lib)] [synthesize:(20:1)] [synthesize:(21:3)] [synthesize:(22:5)] [synthesize:(23:7)] [synthesize:(24:9)] [synthesize:(25:11)] [synth
esize: (70:11)] [synthesize:(70:23)] Unknown variable: (restarg)

**Comment by p0deje** [comment]
Would it be okay if I squash your PR into a single commit?

**Comment by aguspe** [comment]
> Would it be okay if I squash your PR into a single commit?

For sure, that would be perfect, I usually make a lot of small commits to show progress and then I squash everything into one nice commit

**Comment by aguspe** [comment]
> Out of curiosity, what are these warnings?
> 
> > [#type_check_file(lib/selenium/webdriver/remote/http/common.rb@lib)] [synthesize:(20:1)] [synthesize:(21:3)] [synthesize:(22:5)] [synthesize:(23:7)] [synthesize:(24:9)] [synthesize:(25:11)] [synth
> > esize: (70:11)] [synthesize:(70:23)] Unknown variable: (restarg)

From what I researched these errors are related to the splat operator that throws the restarg (unknown variable issue)

I will create this as a separate issue, but before it used to throw an actual error, now it just seems like a warning because of steep itself

https://github.com/soutaro/steep/issues/424

https://github.com/soutaro/steep/issues/7

**Comment by p0deje** [comment]
Thank you again for this massive effort, I'll merge the PR as soon as CI passes.

Out of curiosity, did you use https://github.com/ruby/typeprof or something else to generate RBS files?

**Comment by aguspe** [comment]
> https://github.com/ruby/typeprof

I used RBS prototype: https://github.com/ruby/rbs (This is what rubymine does in the background)

An end user of rbs will probably find rbs prototype the most useful. This command generates boilerplate signature declarations for ruby files. For example, say you have written the below ruby script.

# person.rb
```
class Person
  attr_reader :name
  attr_reader :contacts

  def initialize(name:)
    @name = name
    @contacts = []
  end

  def speak
    "I'm #{@name} and I love Ruby!"
  end
end
```

Rubymine documentation: 

https://blog.jetbrains.com/ruby/2021/09/working-with-rbs-in-rubymine/
https://blog.jetbrains.com/ruby/2021/09/rbs-how-to-get-the-most-out-of-rubymine-code-assistance/

And I was thinking to keep working on this to add proper types now that all the files are there so I can learn more about the selenium ruby library if that is something that will be helpful @p0deje

**Comment by p0deje** [comment]
> And I was thinking to keep working on this to add proper types now that all the files are there so I can learn more about the selenium ruby library if that is something that will be helpful

That would be amazing. I'm curious if you could try typeprof as it might generate more precise signatures, not just boilerplate as rbs. And maybe you won't have to work on manually changing every signature.

**Comment by aguspe** [comment]
> > And I was thinking to keep working on this to add proper types now that all the files are there so I can learn more about the selenium ruby library if that is something that will be helpful
> 
> That would be amazing. I'm curious if you could try typeprof as it might generate more precise signatures, not just boilerplate as rbs. And maybe you won't have to work on manually changing every signature.

I will give it a try to that, thank you for the suggestion 👍
